12-architecture.adoc

```diff
@@ -0,0 +1,133 @@
+
+=== Consistency
+
+Most clients will talk to more than one API. In 2010-15 most organizations getting into API-based architectures might have 1 API, but that started ballooning into double and triple digits with the onset of the microservice trend. 
+
+If your client needs to talk to more than a few APIs, you will probably almost immediately notice inconsistencies between the APIs. Maybe some are internal to your organization and some are external, and that alone means you are interacting with a mixture of GraphQL, REST, gRPC or SOAP.
+
+Maybe you are only talking to internal APIs but they were all built before anyone had a chance to think of a style guide, or your organization just does not care about creating consistent experiences and lets everyone create totally different HTTP authentication approaches, REST hypermedia formats, GraphQL errors are all completely different, and everyone has invented special snowflakes, instead of leveraging API gateways for consistent functionality and https://stoplight.io/open-source/spectral/[enforcing style-guides] with automated tools.
+
+Whatever is going on, talking to multiple APIs probably means dealing with a lot of inconsistency, and without good SDKs that can be a real pain. 
+
+For this, some people use the BFF pattern.
+
+== Backends for Frontends (BFF)
+
+A BFF is an API built for the specific needs of a client. This sounds counterintuitive at first because many of us were taught that APIs must be so generic they can be used by multiple clients, even if those clients require rather different content.
+
+image::images/arch-bff.jpg[]
+
+Trying to make an API so generic that it is useable for every client usually means it is not optimized for any clients. That might be fine depending on how many clients there are, or what they are doing, or how much traffic is coming through the ecosystem, but eventually generic APIs become a source of frustration for the developers who want to build more optimized workflows or data models. 
+
+The worse the design of the remote models the more this starts to become ideal. If you are talking to one of those HTTP/1.1-based RESTish APIs which has jammed 25 resources into a single payload for some reason, but your iPhone application only needs 3 of those resources, making all that extra data go over the wire rather quickly becomes annoying. 
+
+A BFF is a chance for your team to optimize payloads, by grabbing these larger chunks of data and returning only what thinks you probably need. 
```

Comment:
not sure if this is supposed to be 'returning only what it [the BFF] thinks you probably need' or 'returning only what things you probably need'